### DISCUSSION TITLE
Using a Selection to Highlight

### DISCUSSION THREAD

**Discussion by enlore** [uwdata/mosaic#416] [Q&A]
Hey all,

We are building on the examples of Seattle Weather and Normalized Stock Prices examples to create a normalized, highlightable line plot.

```ts
// In case this is useful to see
export const metroTransitColors = [
    "#0039A6",
    "#FF6319",
    "#6CBE45",
    "#996633",
    "#A7A9AC",
    "#FCCC0A",
    "#CE8E00",
    "#EE352E",
    "#00933C",
    "#B933AD",
  ]
```

```tsx
// excerpt showing prop values
<LineFancyPlot
    table={`top_10_growth_cities_normalized`}
    xColumn="job_posted_date"
    yColumn="job_count"
    slicer="slicer"
  />
```

```ts
// function that configures the plot
async function buildViz(
  setVizElement: Dispatch<SetStateAction<Element | null>>,
  table: string,
  xColumn: string | null,
  yColumn: string | null,
  slicer: string | null,
  showTable: boolean,
  yAxisLabel: string,
  xAxisLabel: string,
) {
  const $datePoint = vg.Param.value(new Date("2024-02-01"));
  const $click = vg.Selection.single()
  const $colors = vg.Param.array(metroTransitColors)

  // @ts-expect-error: too many args? nonsense
  const yQuery = vg.sql`
  ${yColumn} / (
    select max(${yColumn})
    from ${table}
    where ${slicer} = source.${slicer}
    and ${xColumn} = ${$datePoint}
  )
`;

  const plotName = `LineFancyPlot-${table}`;

  const linePlot = vg.plot(
    vg.textX({
      x: $datePoint,
      text: $datePoint,
      frameAnchor: "top",
      lineAnchor: "bottom",
      dy: -17,
    }),

    vg.line(vg.from(table), {
      x: xColumn,
      // y: yColumn,
      y: yQuery,
      stroke: slicer,
      strokeWidth: 2.5,
      inset: 5,
      sort: { x: "-y" },
      dx: 10,
    }),

    vg.name(plotName),
    vg.ruleX({ x: $datePoint }),
    vg.colorRange($colors),
    vg.highlight({by: $click}),
    vg.toggle({ 
      as: $click,
      channels: ['x'],
      peers: false
    }),
    vg.nearestX({ as: $datePoint }),

    vg.yTickFormat("%"),
    // vg.yDomain([0, 2]),
    vg.width(1440),
    vg.yGrid(true),
    vg.xTickRotate(45),
    vg.marginBottom(50),
    vg.marginTop(60),
    vg.marginRight(40),
    vg.marginLeft(70),
    vg.xLabel(xAxisLabel),
    vg.yLabel(yAxisLabel),
    vg.style("font-size: 18px"),
    vg.yNice(true),
  );

  const legendElement = vg.colorLegend({
    as: $click,
    for: plotName,
    swatchWidth: 5,
    style: "font-size: 14px",
  });

  if (showTable) {
    setVizElement(
      vg.vconcat(
        vg.hconcat(linePlot, legendElement),
        vg.table({ from: table }),
      ),
    );
  } else {
    setVizElement(vg.vconcat(linePlot, legendElement));
  }
}
```

Generally works quite well, however we are having an issue with the highlighting. Goal is to highlight the relevant line when the user clicks the respective swatch.

See screenshots below for illustration of current behavior.

Thank you in advance.

### No Selection

![Screenshot 2024-05-31 at 2 16 35 PM](https://github.com/uwdata/mosaic/assets/3527094/3124e7ca-1c7d-4602-9f15-cf289ad9e20d)

### Des Moines

![Screenshot 2024-05-31 at 2 16 48 PM](https://github.com/uwdata/mosaic/assets/3527094/fd49fa7e-4929-4182-8b5b-bb548a9d2ee8)

### Anything Else

![Screenshot 2024-05-31 at 2 16 41 PM](https://github.com/uwdata/mosaic/assets/3527094/e6d0e0aa-f3be-4203-b1bd-ea6bc71ae817)

### SVG Opacity Prop

In case this is helpful, I can see in the generated SVG where the opacity prop is being updated.

Des Moines:

![Screenshot 2024-05-31 at 2 58 00 PM](https://github.com/uwdata/mosaic/assets/3527094/d0c0a2b2-de07-48ef-b7ef-7b1a67ce8c4b)

Indianapolis: 

![Screenshot 2024-05-31 at 2 57 48 PM](https://github.com/uwdata/mosaic/assets/3527094/dfbb39d8-4613-4f89-83ff-30d24d6a629c)

### Generated Predicate

![Screenshot 2024-05-31 at 2 31 32 PM](https://github.com/uwdata/mosaic/assets/3527094/9b2196c4-ba78-45c9-b924-7a09f4ab37c0)

### Source Data

[normed_line_top_10_cities.csv](https://github.com/user-attachments/files/15518000/normed_line_top_10_cities.csv)

**Comment by jheer** [comment]
Perhaps try setting the toggle to select properties based on the `stroke` channel rather than the `x` channel? Also try placing the highlight interactor directly after the line mark you wish to highlight? (I wonder if the intervening ruleX mark might get in the way…)

**Reply by jheer** [reply]
I tried this myself but then ran into some bugs in how lines/areas are handled by the `toggle` and `highlight` interactors. So I pushed a fix in ca3f23d, for inclusion in the upcoming v0.9.0 release. Once released, you should be able to select and highlight line series using a toggle interactor over the `z` or `stroke` channels.

**Reply by enlore** [reply]
Confirmed works for our setup via the `stroke` channel.

Edit: I spoke too soon. It highlights different lines now, but it seems like it highlights the wrong line much of the time.

It appears that the state of the $datePoint impacts which line gets highlighted.

When we mouse over the first x value in the series, the expected lines get highlighted.

When we mouse to any other position on the x axis, the incorrect line is usually highlighted.

Screenshots:

### Pointer Over May 2022, Lines Highlight Correctly

#### Columbus

![Screenshot 2024-06-02 at 4 52 03 PM](https://github.com/uwdata/mosaic/assets/3527094/0ae5ae41-e750-47a2-af0a-cb53625c2a3e)

#### Cincinnati

![Screenshot 2024-06-02 at 4 52 00 PM](https://github.com/uwdata/mosaic/assets/3527094/c35bde0f-1532-4061-8aa5-0b69d935cf38)

#### Chicago

![Screenshot 2024-06-02 at 4 51 56 PM](https://github.com/uwdata/mosaic/assets/3527094/39112530-170e-4061-9c9b-3a6c8e646b9d)

### Mouse Over a Different Month/Year

#### Springfield Highlights Cincinnati

![Screenshot 2024-06-02 at 4 48 31 PM](https://github.com/uwdata/mosaic/assets/3527094/b512fda3-f4c4-4abf-975a-1a04f06e567f)

#### Des Moines Highlights Springfield

![Screenshot 2024-06-02 at 4 49 20 PM](https://github.com/uwdata/mosaic/assets/3527094/73c59c0a-28b9-499f-a42c-9adeff2061d9)
